### DISCUSSION TITLE
Airflow Kubernetes Executor multiple namespaces secrets and env variables sharing

### DISCUSSION THREAD

**Discussion by crabio** [apache/airflow#38141] [Q&A]
Hi!
We are using Airflow 2.7.3 deployed in the kubernetes cluster.
We are using KubernetesExecutor to launch pods.

And we are going to launch pods in different namespaces to isolate tasks environments.

Spawning is going well with RoleBinding.
But issue is with secrets sharing.

All pods for tasks requires connection to the database and for this they needs secrets from NS-1 where Airflow Scheduler running.
But in NS-2 we haven't these secrets, because we deploy Airflow in another namespace.

How we can fix it? Maybe we can configure some how pod template to pass all env variables via values from Airflow Sheduler?

**Comment by ArieLevs** [comment]
+1 on the above issue.

Our use case is very similar, just we run airflow tasks in a remote kubernetes cluster, but face similar secrets/configs share between scheduler and worker.

first issue was to provide the remote cluster the gitSync secrets, we ended up creating another secret on the remote cluster, then faced issues with missing fernet key, and so on

what is the best practice here? how should the airflow.cfg reach the remote cluster, or another namespace? using shared storage? is that logical to share sensitive data like fernet via NFS?

* while using kuebernet executor with k8sPodOperator all when super well, since the worker is local and the task is separator to airflow, but with PythonOperator for example, we hit many non trivial issues

**Reply by potiuk** [reply]
We've been disussing this very thing recently, I think one of the ways to achieve it is to share the secrtes you need via K8S pod template and sharing the credentials as env vars via secret created dynamically that could be mounted to the POD run (all that likely possible to be done via custom pod template).

Can you try this @crabio @ArieLevs and see whether this is a viable solution?

One of the drawbacks of this solution is that there is apparently no easy and automated way of removing those secrets when Pods complete, so likely you need to have a separate Job running that will clen-up old unmounted secrets (I believe support for such automated cleaning either had not yet been released in K8S or maybe just recently - but having a separate Job should do the trick and some users implemented similar solutions apparently. 

I think we **might** consider adding support in the future Airflow to do it easier, but for now I think the custom solution like that should be doable and having feedback  from users like you that yes, it's doable, and some feedback on caveats, woudl be a great input to our product disussions.

**Reply by amoghrajesh** [reply]
@crabio @ArieLevs adding to Jarek's point, you can try using it through env variables like this:
```
secrets = [
    {'name': 'my-secret', 'key': 'my-secret-key', 'mount_path': '/etc/secrets/my-secret-key'}
]

task = KubernetesPodOperator(
    task_id='pod_task',
    namespace='my-namespace',
    cmds=['sh', '-c', 'cat /etc/secrets/my-secret-key'],
    secrets=secrets,
    name='airflow-pod',
    dag=dag
)
```

And have a simple CronJob to clean up at some frequency
